smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/commit/ContextHolder.java

```diff
@@ -45,4 +50,18 @@ public void runOnContext(Runnable runnable) {
         }
     }
 
+    public <T> T runOnContextAndAwait(Callable<T> action) {
+        FutureTask<T> task = new FutureTask<>(action);
+        runOnContext(task);
+
+        try {
+            return task.get(timeout, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new WakeupException();
```

Comment:
Are we sure about throwing a WakeupException? It will stop the poll thread dead cold. I used it without giving it much thought on my tests. The other option is to wrap it in a CompletionException along with the others (besides marking the current thread as interrupted). 

Dealing with InterruptedException is always tricky because you never know where they can come from.

Reply:
So, I checked the Apache Kafka client, and it's how it's called:

```java
 private Exception invokePartitionsAssigned(final Set<TopicPartition> assignedPartitions) {
        log.info("Adding newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));

        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
        try {
            final long startMs = time.milliseconds();
            listener.onPartitionsAssigned(assignedPartitions);
            sensors.assignCallbackSensor.record(time.milliseconds() - startMs);
        } catch (WakeupException | InterruptException e) {
            throw e;
        } catch (Exception e) {
            log.error("User provided listener {} failed on invocation of onPartitionsAssigned for partitions {}",
                listener.getClass().getName(), assignedPartitions, e);
            return e;
        }

        return null;
    }
```

The `CompletionException` we are throwing would be handled by the second `catch` block and denotes a problem in the listener. The listener is the source of the issue. So, this case is fine. 

As you mentioned, an interruption can come from anywhere, and it should not be handled by the second block, but by the first one. I believe we should throw an `InterruptException` (note that it's a Kafka-specific unchecked variant of `InterruptedException`) and let the polling thread handle the interruption.

Reply:
Updated the code.